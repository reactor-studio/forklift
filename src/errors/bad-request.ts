import ForkliftError from './forklift-error';

export default class BadRequestError extends ForkliftError {
  constructor(message: string) {
    super(message, 400, 'Bad request');
  }
}
