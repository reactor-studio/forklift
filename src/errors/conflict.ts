import ForkliftError from './forklift-error';

export default class ConflictError extends ForkliftError {
  constructor(message: string) {
    super(message, 409, 'Conflict');
  }
}
