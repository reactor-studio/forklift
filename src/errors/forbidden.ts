import ForkliftError from './forklift-error';

export default class ForbiddenError extends ForkliftError {
  constructor(message: string) {
    super(message, 403, 'Forbidden');
  }
}
