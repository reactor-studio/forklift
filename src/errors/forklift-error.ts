export default class ForkliftError extends Error {
  readonly status: number;

  readonly name: string;

  readonly stack: string;

  constructor(message: string, status: number, name: string) {
    super(message);
    this.message = message;
    this.status = status;
    this.name = name;
    Object.setPrototypeOf(this, new.target.prototype);
    Error.captureStackTrace(this, this.constructor);
  }

  toJson = (showTrace = false): object => {
    return {
      status: this.status,
      title: this.name,
      message: this.message,
      meta: showTrace
        ? {
            trace: this.stack,
          }
        : null,
    };
  };
}
