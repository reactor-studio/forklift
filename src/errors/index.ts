export { default as BadRequestError } from './bad-request';
export { default as ForbiddenError } from './forbidden';
export { default as NotFoundError } from './not-found';
export { default as ConflictError } from './conflict';
export { default as ForkliftError } from './forklift-error';
export { default as InputError } from './input-error';
export { default as OutputError } from './output-error';
