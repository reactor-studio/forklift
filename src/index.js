import IO from './io';

export { IoError, ioErrorAdapter } from './io-error';
export { Status } from './status';
export { setLocals, getLocals } from './locals';
export { IO };
