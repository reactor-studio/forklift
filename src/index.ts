export {
  default as IO,
  Status,
  statusOptions,
  setLocals,
  getLocals,
} from './io';

export { asyncMiddleware, errorMiddleware } from './middleware';

export {
  NotFoundError,
  ForbiddenError,
  InputError,
  OutputError,
  ConflictError,
  ForkliftError,
} from './errors';
