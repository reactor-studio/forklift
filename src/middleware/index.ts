export { asyncMiddleware } from './async-middleware';
export { errorMiddleware } from './error-middleware';
