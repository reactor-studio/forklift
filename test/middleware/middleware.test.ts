import request from 'supertest';
import express from 'express';
import { errorMiddleware, asyncMiddleware } from '../../src/middleware';
import { NotFoundError } from '../../src/errors';

describe('Forklift error middleware', () => {
  it('should handle forklift errors and have meta by default', async () => {
    const testController = {
      get: () => (): Promise<void> => {
        throw new NotFoundError('Hey! Not found!');
      },
    };

    const app = express();
    app.get('/get', testController.get());
    app.use(errorMiddleware());

    const response = await request(app).get('/get');

    expect(response.body.meta).toBeDefined();
    expect(response.body.meta.trace).toBeDefined();
    expect(response.body.title).toBe('Not Found');
  });

  it('should disable meta trace log', async () => {
    const testController = {
      get: () => (): Promise<void> => {
        throw new NotFoundError('Hey! Not found!');
      },
    };

    const app = express();
    app.get('/get', testController.get());
    app.use(errorMiddleware(false));

    const response = await request(app).get('/get');

    expect(response.body.meta).toBeNull();
    expect(response.body).toMatchSnapshot();
    expect(response.body.title).toBe('Not Found');
  });

  it('should handle generic errors gracefully', async () => {
    const testController = {
      get: () => (): Promise<void> => {
        throw new Error('Hey! Not found!');
      },
    };

    const app = express();
    app.get('/get', testController.get());
    app.use(errorMiddleware());

    const response = await request(app).get('/get');

    expect(response.body.stack).toBeDefined();
    expect(response.body.name).toBe('Error');
  });

  it('should use toJSON if available', async () => {
    class CustomError extends Error {
      constructor(msg: string) {
        super(msg);
        Object.setPrototypeOf(this, new.target.prototype);
      }

      toJSON() {
        return { message: 'Yiha' };
      }
    }

    const testController = {
      get: () => (): Promise<void> => {
        throw new CustomError('Hey! Not found!');
      },
    };

    const app = express();
    app.get('/get', testController.get());
    app.use(errorMiddleware());

    const response = await request(app).get('/get');

    expect(response.body.message).toBe('Yiha');
  });

  it('should handle unhandled exceptions in handlers with asyncMiddleware', async () => {
    function sleep(ms: number): Promise<void> {
      return new Promise((resolve) => setTimeout(resolve, ms));
    }

    const testController = {
      get: (): ((
        req: express.Request,
        res: express.Response,
        next: express.NextFunction,
      ) => Promise<void>) =>
        asyncMiddleware(async () => {
          await sleep(100);
          throw new Error('Error');
        }),
    };
    const app = express();
    app.get('/get', testController.get());
    app.use(errorMiddleware(false));

    const response = await request(app).get('/get');

    expect(response.body).toMatchSnapshot();
  });

  it('should call next method if no exception occurred', async () => {
    const nextToCall = jest.fn((_req: any, _res: any, next: any): void => {
      next();
    });

    const testController = {
      get: (): ((
        req: express.Request,
        res: express.Response,
        next: express.NextFunction,
      ) => Promise<void>) =>
        asyncMiddleware(async (_req, _res, next) => {
          expect(next).toBeDefined();
          next();
        }),
    };
    const app = express();
    app.get('/get', testController.get(), nextToCall);
    app.use(errorMiddleware(false));

    await request(app).get('/get');

    expect(nextToCall).toHaveBeenCalled();
  });
});
