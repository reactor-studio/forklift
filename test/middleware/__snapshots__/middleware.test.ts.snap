// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forklift error middleware should disable meta trace log 1`] = `
Object {
  "message": "Hey! Not found!",
  "meta": null,
  "status": 404,
  "title": "Not Found",
}
`;
